core/amber/src/main/scala/edu/uci/ics/texera/web/resource/dashboard/user/dataset/service/GitVersionControlLocalFileStorageService.java

```diff
@@ -0,0 +1,238 @@
+    }
+  }
+
+  /**
+   * Utility class for executing Git commands in a system-agnostic manner.
+   */
+  static class GitSystemCall {
+
+    public static Charset BYTE_2_STRING_CHARSET = StandardCharsets.UTF_8;
+
+    // utility function to execute git-related system call
+    private static byte[] executeGitCommand(String workingDirectory, String... args) throws IOException, InterruptedException {
+      List<String> commands = new ArrayList<>();
+      commands.add("git");
+      Collections.addAll(commands, args);
+
+      ProcessBuilder builder = new ProcessBuilder(commands);
+      builder.directory(new File(workingDirectory));
+      builder.redirectErrorStream(true);
+      Process process = builder.start();
+
+      try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
+        byte[] buffer = new byte[1024];
+        int length;
+        while ((length = process.getInputStream().read(buffer)) != -1) {
+          outputStream.write(buffer, 0, length);
+        }
+
+        int exitCode = process.waitFor();
+        if (exitCode != 0) {
+          throw new IOException("Failed to execute Git command: " + String.join(" ", commands));
+        }
+
+        return outputStream.toByteArray();
+      }
+    }
+
+    public static void initRepo(String path) throws IOException, InterruptedException {
+      executeGitCommand(path, "init");
+    }
+
+    // this function adds all changes to the stage and create a commit, it will return the hash of that commit
+    public static String addAndCommit(String repoPath, String commitMessage) throws IOException, InterruptedException {
+      // Adding all files and committing
+      executeGitCommand(repoPath, "add", ".");
+      executeGitCommand(repoPath, "commit", "-m", commitMessage);
+
+      // Retrieving the full commit hash of the latest commit
+      byte[] commitHashOutput = executeGitCommand(repoPath, "rev-parse", "HEAD");
+      return new String(commitHashOutput, StandardCharsets.UTF_8).trim();
+    }
+
+    // write the file content of a certain commit into the output file stream
+    public static void showFileContentOfCommit(String repoPath, String commitHash, String filePath, OutputStream outputStream) throws IOException, InterruptedException {
+      byte[] fileContent = executeGitCommand(repoPath, "show", commitHash + ":" + filePath);
+      outputStream.write(fileContent);
+    }
+
+    // get the file tree of a certain commit as a Map
+    public static Map<String, Object> getFileTreeHierarchy(String repoPath, String commitHash) throws IOException, InterruptedException {
+      String treeOutput = new String(executeGitCommand(repoPath, "ls-tree", "-r", commitHash), BYTE_2_STRING_CHARSET);
+      return parseFileTree(treeOutput);
+    }
+
+    // utility function to parse the file tree as the map
+    private static Map<String, Object> parseFileTree(String treeOutput) {
+      Map<String, Object> fileTree = new HashMap<>();
```

Comment:
it is generally not a good idea to invent your own parser of a specific format. Does git provide other formats of the same information, for example, do they provide JSON or XML?

Reply:
A new class, FileNode, is introduced.